### PR TITLE
Issue #465: Add job duration metric for monitor jobs

### PR DESCRIPTION
**Pull request by MedMaalej**
* Add a method that sets the duration metric in AbstractStrategy
* Call the created method
* Update unit tests
* Initialize and refresh metrics after discovery job in MonitoringTask

### CODE REVIEW

**Comment by MedMaalej** [comment]
I tested this ticket using windows connectors locally.

The job duration metric was displayed correctly in MetricsHub logs.

Here are the screenshots of the Grafana Dashboard that uses MetricsHub metrics:
Capture d'écran 2024-12-02 134638

Capture d'écran 2024-12-02 134657

Capture d'écran 2024-12-02 134706

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/AfterAllStrategyTest.java [hunk 1]
```diff
@@ -264,6 +264,41 @@ void testRunFromCollect() {
...
+				)
+				.getValue()
+		);
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"afterAll\"," + " monitor.type=\"none\", connector_id=\"afterAllSource\"}"
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/AfterAllStrategyTest.java:287]: It is not required to concatenate strings.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/simple/SimpleStrategyTest.java [hunk 2]
```diff
@@ -228,5 +229,29 @@ void testRun() throws Exception {
...
 		);
+
+		// Check job duration metrics values
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("monitor1")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"simple\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/simple/SimpleStrategyTest.java:240]: String concat is not required.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/simple/SimpleStrategyTest.java [hunk 3]
```diff
@@ -228,5 +229,29 @@ void testRun() throws Exception {
...
+				)
+				.getValue()
+		);
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("monitor1")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"simple\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/simple/SimpleStrategyTest.java:251]: Same as previous comment about string concat.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/AfterAllStrategyTest.java [hunk 4]
```diff
@@ -264,6 +264,41 @@ void testRunFromCollect() {
...
 		assertEquals("health-ok", sourceTableLine.get(9));
+
+		// Check job duration metrics values
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"collect\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/AfterAllStrategyTest.java:275]: Same as previous comments about string concatenation.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java [hunk 5]
```diff
@@ -264,6 +264,42 @@ void testRunFromCollect() {
...
 		assertEquals("health-ok", sourceTableLine.get(9));
+
+		// Check job duration metrics values
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"collect\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java:275]: Update please.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java [hunk 6]
```diff
@@ -264,6 +264,42 @@ void testRunFromCollect() {
...
+				)
+				.getValue()
+		);
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"beforeAll\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java:287]: Update please.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java [hunk 7]
```diff
@@ -264,6 +264,42 @@ void testRunFromCollect() {
...
+				)
+				.getValue()
+		);
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"collect\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/surrounding/BeforeAllStrategyTest.java:298]: Update please.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/CollectStrategyTest.java [hunk 8]
```diff
@@ -259,5 +260,29 @@ void testRun() throws Exception {
...
 		);
+
+		// Check job duration metrics values
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"collect\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/CollectStrategyTest.java:271]: It not required to concatenate strings.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/CollectStrategyTest.java [hunk 9]
```diff
@@ -259,5 +260,29 @@ void testRun() throws Exception {
...
+				)
+				.getValue()
+		);
+		assertNotNull(
+			telemetryManager
+				.getMonitors()
+				.get("host")
+				.get("anyMonitorId")
+				.getMetric(
+					"metricshub.job.duration{job.type=\"collect\"," +
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/CollectStrategyTest.java:282]: do not concatenate strings.

**Review by NassimBtk — CHANGES_REQUESTED** [review]
Updates requires on junit tests.

File: metricshub-agent/src/main/java/org/sentrysoftware/metricshub/agent/service/task/MonitoringTask.java [hunk 10]
```diff
@@ -117,6 +117,12 @@ public void run() {
 			// Initialize the OpenTelemetry observers and LogEmitter after the discovery
 			// as at this time we should have what we want to observe
 			initOtelSdk(telemetryManager, resourceConfig);
+
+			// Initialize metric observers
```
Comment (NassimBtk) [inline metricshub-agent/src/main/java/org/sentrysoftware/metricshub/agent/service/task/MonitoringTask.java:121]: This requires an additional comment to explain why we are doing this here, please.

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/AbstractStrategy.java [hunk 11]
```diff
@@ -506,4 +506,40 @@ public boolean isMonitorFiltered(final String monitorType) {
...
+			.append("\", monitor.type=\"")
+			.append(monitorType)
+			.append("\", connector_id=\"")
+			.append(connectorId)
+			.append("\"}")
+			.toString();
+		metricFactory.collectNumberMetric(
+			endpointHostMonitor,
+			jobDurationMetricKey,
+			(double) (endTime - startTime) / 1000, // Job duration in seconds
```
Comment (NassimBtk) [inline metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/AbstractStrategy.java:541]:
```suggestion
			(endTime - startTime) / 1000.0, // Job duration in seconds
```

**Review by NassimBtk — CHANGES_REQUESTED** [review]

**Review by NassimBtk — APPROVED** [review]

**Pull request merged**